lib/context.js

```diff
@@ -1041,19 +1041,107 @@ function _expandIri(activeCtx, value, relativeTo, localCtx, defined, options) {
     }
   }
 
-  // prepend vocab
   if(relativeTo.vocab && '@vocab' in activeCtx) {
-    return activeCtx['@vocab'] + value;
-  }
+    // prepend vocab
+    const prependedResult = activeCtx['@vocab'] + value;
+    let expansionMapResult = undefined;
+    if(options && options.expansionMap) {
+      // if we are about to expand the value by prepending
+      // @vocab then call the expansion map to inform
+      // interested callers that this is occurring
+
+      // TODO: use `await` to support async
+      expansionMapResult = options.expansionMap({
+        prependedIri: {
+          type: '@vocab',
+          vocab: activeCtx['@vocab'],
+          value,
+          result: prependedResult
+        },
+        activeCtx,
+        options
+      });
 
-  // prepend base
-  if(relativeTo.base && '@base' in activeCtx) {
+    }
+    if(expansionMapResult !== undefined) {
+      value = expansionMapResult;
+    } else {
+      // the null case preserves value as potentially relative
+      value = prependedResult;
+    }
+  } else if(relativeTo.base && '@base' in activeCtx) {
+    // prepend base
     if(activeCtx['@base']) {
```

Comment:
I think we can simplify the logic here and consolidate the new code into one place with a single conditional on just `relativeTo.base` at the top-level ... and then getting the base value from `@base in activeCtx` and falling back to `options.base`. That would also make sure we still use the `expansionMap` when `activeCtx['@base']` is `null` ... because I think we may be failing to do so here at the moment.

@davidlehn -- any feedback on what we're passing the `expansionMap` handler here?

Reply:
@dlongley thanks I agree, have updated the PR accordingly.